polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -552,6 +548,16 @@ Value MLIRScanner::castToMemSpace(Value Val, unsigned MemSpace) {
         if (ValType.getMemorySpaceAsInt() == MemSpace)
           return Val;
 
+        if (GenerateSYCLAddrSpaceCast)
+          return Builder.create<sycl::SYCLAddrSpaceCastOp>(
+              Loc,
+              MemRefType::get(ValType.getShape(), ValType.getElementType(),
+                              ValType.getLayout(),
+                              mlirclang::wrapIntegerMemorySpace(
+                                  MemSpace, ValType.getContext())),
+              Val);
```

Comment:
We're still allowing the operation to be created in a non-SYCL context. I suggest dropping the option and checking the insertion context (`Host` or `SYCLDevice`) here.

Reply:
In a non-SYCL context, `GenerateSYCLAddrSpaceCast` must be `false`.
```
  if (SYCLIsDevice) {
    Ctx.getOrLoadDialect<mlir::sycl::SYCLDialect>();
    // TODO: Use memref.memory_space_cast by default.
    if (GenerateSYCLAddrSpaceCast.getNumOccurrences() == 0)
      GenerateSYCLAddrSpaceCast = true;
  } else if (GenerateSYCLAddrSpaceCast) {
    CGEIST_WARNING(
        llvm::WithColor::warning()
        << "Cannot use sycl.addrspacecast outside of SYCL context\n");
    GenerateSYCLAddrSpaceCast = false;
  }
```

Reply:
Since we are going to generate `memref.memory_space_cast`, I would like to avoid modify multiple function signatures to pass the context. What do you think?